### PR TITLE
fix: update Docker single-node cluster documentation

### PR DESCRIPTION
**Pull request by sudomateo**
The documentation for running a single-node cluster locally with Docker was incorrect for the following reasons:

- The `docker run` command was passing `--net=roachnet` without telling the user how to create the `roachnet` network. Additionally, the `roachnet` network is not needed when deploying a single-node cluster.
- The `docker run` command was passing `--http-addr=localhost:8080` which made the DB Console listen on `localhost` where users could not connect to it from their host machine.
- The command to connect to the cluster using the built-in SQL client was using the incorrect hostname/IP, leading to x509 certificate errors when trying to connect to the DB Console (after fixing the `--http-addr` issue above).

The documentation has been updated to omit unnecessary options. A new section was also added to encourage a user to connect to the DB Console whereas it was previously easy to miss when working through the documentation. Lastly, the `docker log` command was updated to `docker logs`, the correct command.

### CODE REVIEW

**Comment by taroface** [comment]
cc @mdlinville

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md [hunk 1]
```diff
@@ -105,21 +103,25 @@ The `cockroach` process listens on `127.0.0.1:26257` and `localhost:26257`, and
...
+    docker exec -it roach-single ./cockroach sql --url="postgresql://root@127.0.0.1:26257/defaultdb?sslcert=certs%2Fclient.root.crt&sslkey=certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=certs%2Fca.crt"
     ~~~
 
-1. To connect to the cluster using DB Console in your web browser, navigate to the address printed next to `webui:` in the cluster's startup details. You can replace the IP address with the hostname of the Docker container. To connect to the `roach-single` cluster, navigate to [https://roach-single:8080/](https://roach-single:8080/) for a secure cluster or [http://roach-single:8080/](http://roach-single:8080/) for an insecure cluster.
+### Step 4. Access the DB Console
 
-### Step 4. Stop the cluster
+The [DB Console]({% link {{ page.version.version }}/ui-overview.md %}) gives you insight into the overall health of your cluster as well as the performance of the client workload.
+
+1. When you started the node's container, you mapped the node's default HTTP port `8080` to port `8080` on the Docker host, so go to http://localhost:8080. If necessary, replace `localhost` with the hostname or IP address of the Docker host.
```
Comment (mdlinville) [inline src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:122]:
```suggestion
When you started the node's container, you mapped the node's default HTTP port `8080` to port `8080` on the Docker host, and you published port 8080 so that it can be accessed from outside the Docker host. To connect to DB Console, go to `http://localhost:8080`. If necessary, replace `localhost` with the hostname or IP address of the Docker host.
```

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md [hunk 2]
```diff
@@ -59,12 +59,10 @@ The `cockroach` process listens on `127.0.0.1:26257` and `localhost:26257`, and
       --env COCKROACH_PASSWORD={PASSWORD} \
       --name=roach-single \
       --hostname=roach-single \
-      --net=roachnet \
```
Comment (mdlinville) [inline src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:62]: Thanks, leaving the network details in the single-node example was a mistake on my part.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md [hunk 3]
```diff
@@ -59,12 +59,10 @@ The `cockroach` process listens on `127.0.0.1:26257` and `localhost:26257`, and
...
       -p 26257:26257 \
       -p 8080:8080 \
       -v "roach-single:/cockroach/cockroach-data"  \
-      {{page.release_info.docker_image}}:{{page.release_info.version}} start-single-node \
-      --http-addr=localhost:8080 \
+      {{page.release_info.docker_image}}:{{page.release_info.version}} start-single-node
```
Comment (mdlinville) [inline src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:65]: I prefer for you to revert this line, because we should specify where DB Console should listen. The `--listen-addr` flag is ignored for single-node deployments, but `--http-addr` is not, if I understand correctly.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md [hunk 4]
```diff
@@ -105,21 +103,25 @@ The `cockroach` process listens on `127.0.0.1:26257` and `localhost:26257`, and
...
     ~~~ shell
-    docker log roach-single --follow
+    docker logs --follow roach-single
     ~~~
 
-    To stop monitoring the logs, press Ctrl+C to exit the `docker log` command.
+    To stop monitoring the logs, press Ctrl+C to exit the `docker logs` command.
 
-1. To connect to the cluster interactively using the `cockroach sql` command-line interface, set `--url` cluster's SQL connection string, which is printed next to `sql:` in the cluster's startup details. You can replace the IP address with the hostname of the Docker container so that the script continues to work if the IP address changes. To run `cockroach sql` on the `roach-single` cluster:
+1. To connect to the cluster interactively using the `cockroach sql` command-line interface, set `--url` cluster's SQL connection string, which is printed next to `sql:` in the cluster's startup details. To run `cockroach sql` on the `roach-single` cluster:
```
Comment (mdlinville) [inline src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:111]:
```suggestion
1. To connect to the cluster interactively using the `cockroach sql` command-line interface, set `--url` cluster's SQL connection string, which is printed next to `sql:` in the cluster's startup details. Connect to the `roach-single` cluster:
```

**Review by mdlinville — COMMENTED** [review]
This looks great. I have a few suggestions. Also, it would be great if you could update the non-single-node section title [here](https://www.cockroachlabs.com/docs/v23.1/start-a-local-cluster-in-docker-linux#step-4-use-the-built-in-sql-client) to "Connect to the cluster" also. The include is in the same directory as the one you're editing here.

My feedback also applies to the 23.2 page.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 2 above)
Comment (sudomateo) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:62]: You're welcome!

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 3 above)
Comment (sudomateo) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:65]: When 
```
$ podman run -it --rm --network container:roach-single ubuntu:latest bash -c "apt update && apt install -y iproute2 && ss -plnt"
...
State          Recv-Q         Send-Q                  Local Address:Port                    Peer Address:Port         Process
LISTEN         0              0                           127.0.0.1:8080                         0.0.0.0:*
LISTEN         0              0                                   *:26257                              *:*
```

```

State          Recv-Q         Send-Q                  Local Address:Port                    Peer Address:Port         Process
LISTEN         0              0                                   *:8080                               *:*
LISTEN         0              0                                   *:26257                              *:*
```

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 1 above)
Comment (sudomateo) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:122]: Love it, implemented!

**Comment by sudomateo** [comment]
Thank you for the review @mdlinville! I implemented the requested feedback and left a comment on the conversation to revert the `--http-addr` removal. Let me know what you think.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 3 above)
Comment (mdlinville) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:65]: Looks like `http-address` is also ignored in single mode. @ZhouXing19 Can you weigh in here? I believe you did some work on this previously: https://github.com/cockroachdb/cockroach/pull/70238

Comment (mdlinville) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:65]: Jane is going to look at this tomorrow, but from looking at https://github.com/ZhouXing19/cockroach/blob/master/build/deploy/cockroach.sh it looks like you can specify a different advertise address but if you specify a different advertise port, we throw it away and stubbornly use 8080.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 4 above)
Comment (mdlinville) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md]: I'm not sure this suggestion got applied -- maybe you haven't pushed your latest changes yet.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 1 above)
Comment (mdlinville) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md]: I don't think this change has been applied yet in your PR -- maybe you need to push your latest changes.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 4 above)
Comment (mdlinville) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md]: Maybe I just needed to refresh.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 1 above)
Comment (mdlinville) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md]: Maybe I just needed to refresh. Sorry about that.

Comment (sudomateo) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md]: Sorry! I didn't use the UI to implement your suggestions. I implemented them as part of https://github.com/cockroachdb/docs/pull/18045/commits/12566040e7d99670bda1187c0f985dbd71affcca.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 4 above)
Comment (sudomateo) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md]: Sorry! I didn't use the UI to implement your suggestions. I implemented them as part of https://github.com/cockroachdb/docs/pull/18045/commits/12566040e7d99670bda1187c0f985dbd71affcca.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 3 above)
Comment (sudomateo) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:65]: Thank you for the spelunking! I'll wait for Jane to weigh in so we can discuss this further.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md [hunk 5]
```diff
@@ -119,7 +119,7 @@ The `cockroach` process listens on `127.0.0.1:26257` and `localhost:26257`, and
 
 The [DB Console]({% link {{ page.version.version }}/ui-overview.md %}) gives you insight into the overall health of your cluster as well as the performance of the client workload.
 
-1. When you started the node's container, you mapped the node's default HTTP port `8080` to port `8080` on the Docker host, so go to http://localhost:8080. If necessary, replace `localhost` with the hostname or IP address of the Docker host.
+When you started the node's container, you mapped the node's default HTTP port `8080` to port `8080` on the Docker host, and you published port 8080 so that it can be accessed from outside the Docker host. To connect to DB Console, go to `http://localhost:8080`. If necessary, replace `localhost` with the hostname or IP address of the Docker host.
```
Comment (mdlinville) [inline src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:122]:
```suggestion
When you started the node's container, you mapped the node's default HTTP port `8080` to port `8080` on the Docker host, and you published port 8080 so that it can be accessed from outside the Docker host. To connect to the DB Console, go to `http://localhost:8080`. If necessary, replace `localhost` with the hostname or IP address of the Docker host.
```

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps.md [hunk 6]
```diff
@@ -170,7 +170,7 @@ When SQL and inter-node traffic are separated, some client commands need to be m
     nodeID:              1
     ~~~
 
-### Step 4. Use the built-in SQL client
+### Step 4. Connect to the cluster
```
Comment (mdlinville) [inline src/current/_includes/v23.1/start-in-docker/mac-linux-steps.md:173]: <3

File: src/current/_includes/v23.2/start-in-docker/mac-linux-steps-single-node.md [hunk 7]
```diff
@@ -119,7 +119,7 @@ The `cockroach` process listens on `127.0.0.1:26257` and `localhost:26257`, and
 
 The [DB Console]({% link {{ page.version.version }}/ui-overview.md %}) gives you insight into the overall health of your cluster as well as the performance of the client workload.
 
-1. When you started the node's container, you mapped the node's default HTTP port `8080` to port `8080` on the Docker host, so go to http://localhost:8080. If necessary, replace `localhost` with the hostname or IP address of the Docker host.
+When you started the node's container, you mapped the node's default HTTP port `8080` to port `8080` on the Docker host, and you published port 8080 so that it can be accessed from outside the Docker host. To connect to DB Console, go to `http://localhost:8080`. If necessary, replace `localhost` with the hostname or IP address of the Docker host.
```
Comment (mdlinville) [inline src/current/_includes/v23.2/start-in-docker/mac-linux-steps-single-node.md:122]:
```suggestion
When you started the node's container, you mapped the node's default HTTP port `8080` to port `8080` on the Docker host, and you published port 8080 so that it can be accessed from outside the Docker host. To connect to the DB Console, go to `http://localhost:8080`. If necessary, replace `localhost` with the hostname or IP address of the Docker host.
```

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 3 above)
Comment (ZhouXing19) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:65]: Thanks for reporting this issue. In the current implementation `http-address` should be resolved and applied. 
But I'm not 100% sure if I understand the problem here correctly, and my following test seems to confirm my idea:

```
❯ docker run -d --name=roach5 --hostname=roach5 -p 26258:26257 -p 8081:8080   cockroachdb/cockroach-local:latest start-single-node --insecure --http-addr=localhost:8081
0c19b937246131ce1bcc26551f59cfd24f5b212a238336b9fec45a85cabd8dfa

❯ nc -vz localhost 26257         
nc: connectx to localhost port 26257 (tcp) failed: Connection refused
nc: connectx to localhost port 26257 (tcp) failed: Connection refused


❯ nc -vz localhost 26258
nc: connectx to localhost port 26258 (tcp) failed: Connection refused
Connection to localhost port 26258 [tcp/*] succeeded!


❯ nc -vz localhost 8080 
nc: connectx to localhost port 8080 (tcp) failed: Connection refused
nc: connectx to localhost port 8080 (tcp) failed: Connection refused

❯ nc -vz localhost 8081
nc: connectx to localhost port 8081 (tcp) failed: Connection refused
Connection to localhost port 8081 [tcp/sunproxyadmin] succeeded!
```

So it seems that we did exposed the http port via 8081, and it is accessible.

Comment (sudomateo) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:65]: Thank you for getting back to us. I believe the `nc` test is hiding the issue since `nc`'s `-z` option will only report the connection status without establishing a connection. If you remove the `-z` option you'll see that the TCP connection gets reset and control is returned immediately back to your terminal. This connection reset can be seen using `curl` as well.

```sh
$ docker run -d --name=roach-single --hostname=roach-single -p 3000:26257 -p 4000:8080 docker.io/cockroachdb/cockroach:v23.1.11 start-single-node --insecure --http-addr=localhost:8080
f0ab87b54fd5fb18bb5275bc8c97aa2db099ef11add7e3e1bb7570d5771f943c

$ nc -v localhost 4000
Connection to localhost port 4000 [tcp/terabase] succeeded!

$ curl -vL http://localhost:4000
*   Trying 127.0.0.1:4000...
* Connected to localhost (127.0.0.1) port 4000 (#0)
> GET / HTTP/1.1
> Host: localhost:4000
> User-Agent: curl/8.1.2
> Accept: */*
>
* Recv failure: Connection reset by peer
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
```

I believe this is because the host's `4000` binding cannot communicate with the container's `8080` binding since the container's binding is listening on `localhost`, not `*`.

If we remove the `--http-addr` option, the container's `8080` binding will listen on `*` instead of `localhost`, and the host's `4000` binding can communicate with the container's binding. With this change, I can now use `nc` to make a raw HTTP request and then quit with Control-D. Using `curl` also now works.

```sh
$ docker run -d --name=roach-single --hostname=roach-single -p 3000:26257 -p 4000:8080 docker.io/cockroachdb/cockroach:v23.1.11 start-single-node --insecure --http-addr=localhost:8080
2bf0084bcd511cd2df59cccecd97e85c60423da93d3b0a0bb38ecee24763e74b

> nc -v localhost 4000
Connection to localhost port 4000 [tcp/terabase] succeeded!
GET / HTTP/1.1
Host: localhost:4000

HTTP/1.1 200 OK
Accept-Ranges: bytes
Cache-Control: no-cache
Content-Length: 264
Content-Type: text/html; charset=utf-8
Last-Modified: Wed, 27 Sep 2023 02:36:23 GMT
Vary: Accept-Encoding
Date: Thu, 09 Nov 2023 22:58:39 GMT

<!DOCTYPE html>
<html>
        <head>
                <title>Cockroach Console</title>
                <meta charset="UTF-8">
                <link href="favicon.ico" rel="shortcut icon">
        </head>
        <body>
                <div id="react-layout"></div>
                <script src="bundle.js" type="text/javascript"></script>
        </body>
</html>
^D

$ curl -vL http://localhost:4000
*   Trying 127.0.0.1:4000...
* Connected to localhost (127.0.0.1) port 4000 (#0)
> GET / HTTP/1.1
> Host: localhost:4000
> User-Agent: curl/8.1.2
> Accept: */*
>
< HTTP/1.1 200 OK
< Accept-Ranges: bytes
< Cache-Control: no-cache
< Content-Length: 264
< Content-Type: text/html; charset=utf-8
< Last-Modified: Wed, 27 Sep 2023 02:36:23 GMT
< Vary: Accept-Encoding
< Date: Thu, 09 Nov 2023 22:55:20 GMT
<
<!DOCTYPE html>
<html>
        <head>
                <title>Cockroach Console</title>
                <meta charset="UTF-8">
                <link href="favicon.ico" rel="shortcut icon">
        </head>
        <body>
                <div id="react-layout"></div>
                <script src="bundle.js" type="text/javascript"></script>
        </body>
</html>
* Connection #0 to host localhost left intact
```

Comment (ZhouXing19) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:65]: Thanks for detailed explanation. Yes, it makes sense to me to not use `--http-addr=localhost:8080`. However, `--http-addr=roach-single:8080` or `--http-addr=0.0.0.0:8080` should work. 

```
❯ docker run -d --name=roach-single --hostname=roach-single -p 3000:26257 -p 4000:8081 cockroachdb/cockroach:latest-v23.1 start-single-node --insecure --http-addr=roach-single:8081 
1e9ead570cf6978e66d7d3addc00602338f38105fcfd0e4d38e0096be13bfb57

❯ curl -vL http://127.0.0.1:4000
*   Trying 127.0.0.1:4000...
* Connected to 127.0.0.1 (127.0.0.1) port 4000 (#0)
> GET / HTTP/1.1
> Host: 127.0.0.1:4000
> User-Agent: curl/7.84.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Accept-Ranges: bytes
< Cache-Control: no-cache
< Content-Length: 264
< Content-Type: text/html; charset=utf-8
< Last-Modified: Wed, 27 Sep 2023 01:53:43 GMT
< Vary: Accept-Encoding
< Date: Mon, 13 Nov 2023 18:52:07 GMT
< 
<!DOCTYPE html>
<html>
        <head>
                <title>Cockroach Console</title>
                <meta charset="UTF-8">
                <link href="favicon.ico" rel="shortcut icon">
        </head>
        <body>
                <div id="react-layout"></div>
                <script src="bundle.js" type="text/javascript"></script>
        </body>
</html>
* Connection #0 to host 127.0.0.1 left intact
```

Essentially, IIUC the key here is that though`--http-addr` flag is _not_ ignored for the `start-single-node` command, the port forwarding supported by docker always maps a TCP port inside a container, rather than an ip-specific port. The docker doc I found didn't explicitly point this out, but it seems to be [a commonly recommended practice](https://stackoverflow.com/a/57427805/22315968) to let the app inside the docker bind itself to 0.0.0.0 rather than 127.0.0.1.

Comment (sudomateo) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:65]: Thank you for confirming! Given this information, I think it's best to use `--http-addr=roach-single:8080` in the documentation.

@mdlinville what do you think?

**Review by mdlinville — APPROVED** [review]

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md [hunk 8]
```diff
@@ -105,21 +104,25 @@ The `cockroach` process listens on `127.0.0.1:26257` and `localhost:26257`, and
...
 
-1. To connect to the cluster interactively using the `cockroach sql` command-line interface, set `--url` cluster's SQL connection string, which is printed next to `sql:` in the cluster's startup details. You can replace the IP address with the hostname of the Docker container so that the script continues to work if the IP address changes. To run `cockroach sql` on the `roach-single` cluster:
+1. To connect to the cluster interactively using the `cockroach sql` command-line interface, set `--url` cluster's SQL connection string, which is printed next to `sql:` in the cluster's startup details. Connect to the `roach-single` cluster:
 
     {% include_cached copy-clipboard.html %}
     ~~~ shell
-    docker exec -it roach-single ./cockroach sql --url="postgresql://root@roach-single:26257/defaultdb?sslcert=certs%2Fclient.root.crt&sslkey=certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=certs%2Fca.crt"
+    docker exec -it roach-single ./cockroach sql --url="postgresql://root@127.0.0.1:26257/defaultdb?sslcert=certs%2Fclient.root.crt&sslkey=certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=certs%2Fca.crt"
```
Comment (mdlinville) [inline src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:116]: Can you revert this change after recent discussion?

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md [hunk 9]
```diff
@@ -105,21 +104,25 @@ The `cockroach` process listens on `127.0.0.1:26257` and `localhost:26257`, and
...
     ~~~
 
-1. To connect to the cluster using DB Console in your web browser, navigate to the address printed next to `webui:` in the cluster's startup details. You can replace the IP address with the hostname of the Docker container. To connect to the `roach-single` cluster, navigate to [https://roach-single:8080/](https://roach-single:8080/) for a secure cluster or [http://roach-single:8080/](http://roach-single:8080/) for an insecure cluster.
+### Step 4. Access the DB Console
 
-### Step 4. Stop the cluster
+The [DB Console]({% link {{ page.version.version }}/ui-overview.md %}) gives you insight into the overall health of your cluster as well as the performance of the client workload.
+
+When you started the node's container, you mapped the node's default HTTP port `8080` to port `8080` on the Docker host, and you published port 8080 so that it can be accessed from outside the Docker host. To connect to the DB Console, go to `http://localhost:8080`. If necessary, replace `localhost` with the hostname or IP address of the Docker host.
```
Comment (mdlinville) [inline src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:123]: Can you provide a bit more explanation here to reflect Jane's explanation?

File: src/current/_includes/v23.2/start-in-docker/mac-linux-steps-single-node.md [hunk 10]
```diff
@@ -105,21 +104,25 @@ The `cockroach` process listens on `127.0.0.1:26257` and `localhost:26257`, and
...
 
-1. To connect to the cluster interactively using the `cockroach sql` command-line interface, set `--url` cluster's SQL connection string, which is printed next to `sql:` in the cluster's startup details. You can replace the IP address with the hostname of the Docker container so that the script continues to work if the IP address changes. To run `cockroach sql` on the `roach-single` cluster:
+1. To connect to the cluster interactively using the `cockroach sql` command-line interface, set `--url` cluster's SQL connection string, which is printed next to `sql:` in the cluster's startup details. Connect to the `roach-single` cluster:
 
     {% include_cached copy-clipboard.html %}
     ~~~ shell
-    docker exec -it roach-single ./cockroach sql --url="postgresql://root@roach-single:26257/defaultdb?sslcert=certs%2Fclient.root.crt&sslkey=certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=certs%2Fca.crt"
+    docker exec -it roach-single ./cockroach sql --url="postgresql://root@127.0.0.1:26257/defaultdb?sslcert=certs%2Fclient.root.crt&sslkey=certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=certs%2Fca.crt"
```
Comment (mdlinville) [inline src/current/_includes/v23.2/start-in-docker/mac-linux-steps-single-node.md:116]: Can you revert this change after recent discussion?

**Review by mdlinville — CHANGES_REQUESTED** [review]
I approved and then noticed a couple of areas that could use a bit more work. Sorry about the rebase -- I really thought my next action was going to be to merge. Thanks for your patience.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 8 above)
Comment (sudomateo) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:116]: This change needs to remain. Otherwise users will receive an x509 error when trying to connect to the database:

```sh
$ docker exec -it roach-single ./cockroach sql --url="postgresql://root@roach-single:26257/defaultdb?sslcert=certs%2Fclient.root.crt&sslkey=certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=certs%2Fca.crt"
#
# Welcome to the CockroachDB SQL shell.
# All statements must be terminated by a semicolon.
# To exit, type: \q.
#
ERROR: failed to connect to `host=roach-single user=root database=defaultdb`: failed to write startup message (x509: certificate is not valid for any names, but wanted to match roach-single)
Failed running "sql"
```

File: src/current/_includes/v23.2/start-in-docker/mac-linux-steps-single-node.md (hunk 10 above)
Comment (sudomateo) [reply src/current/_includes/v23.2/start-in-docker/mac-linux-steps-single-node.md:116]: Responded in the other thread that this change needs to remain.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 9 above)
Comment (sudomateo) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md]: Updated this to better explain what's happening.

File: src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md (hunk 8 above)
Comment (mdlinville) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:116]: Oh man, I just noticed something about the `start-single-node` procedure. It's not using `--insecure`. I guess maybe that is disallowed for a single-node cluster. We can put that to the side for now. It would be better to _eventually_ (outside of this PR) update the earlier procedure to not use `--insecure` I think, though it introduces an extra complication.

I further think that the problem here is that the certificate inside the container is generated only for `localhost` and not for the container's name (which is also its DNS hostname), so you're seeing the x509 error. I think this points to the potential for a feature request to create the certificate with all of the in-container IPs it is binding to, and their resolved DNS hostnames. WDYT? cc/ @ZhouXing19

Comment (sudomateo) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:116]: > Oh man, I just noticed something about the `start-single-node` procedure. It's not using `--insecure`. I guess maybe that is disallowed for a single-node cluster. We can put that to the side for now. It would be better to _eventually_ (outside of this PR) update the earlier procedure to not use `--insecure` I think, though it introduces an extra complication.

It's not disallowed. I actually had my changes using `--insecure` originally until I saw that the page mentioned "By default, a `certs` directory is created and CockroachDB starts in secure mode." right after the `docker run` command. I figured that it was intentionally starting the cluster in secure mode even though the multi-node cluster instructions above it used `--insecure`. Happy to introduce `--insecure` if it's desired here.

> I further think that the problem here is that the node certificate inside the container is generated only for `localhost` and not for the container's name (which is also its DNS hostname), so you're seeing the x509 error. I think this points to the potential for a feature request to create the certificate with all of the in-container IPs it is binding to, and their resolved DNS hostnames. WDYT? cc/ @ZhouXing19

The TLS certificate that's generated uses `node` as the Common Name (CN) and `127.0.0.1` as IP Subject Alternative Names (SANs). Trying to connect using the `roach-single` name will give x509 errors.

```sh
$ docker cp roach-single:/cockroach/certs/node.crt ./node.pem && openssl x509 -in ./node.pem -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            5d:65:09:26:a9:92:8a:b8:cc:73:3a:7d:70:af:50:07
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: O = Cockroach, CN = Cockroach CA
        Validity
            Not Before: Nov 14 22:19:55 2023 GMT
            Not After : Nov 18 22:19:55 2028 GMT
        Subject: O = Cockroach, CN = node
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    REDACTED
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage:
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 Authority Key Identifier:
                BD:43:1A:9A:9D:06:CC:6A:73:77:25:1B:69:85:B6:7B:45:F1:A2:59
            X509v3 Subject Alternative Name:
                IP Address:127.0.0.1, IP Address:127.0.0.1
    Signature Algorithm: sha256WithRSAEncryption
    Signature Value:
        REDACTED
-----BEGIN CERTIFICATE-----
MIIDQTCCAimgAwIBAgIQXWUJJqmSirjMczp9cK9QBzANBgkqhkiG9w0BAQsFADAr
MRIwEAYDVQQKEwlDb2Nrcm9hY2gxFTATBgNVBAMTDENvY2tyb2FjaCBDQTAeFw0y
MzExMTQyMjE5NTVaFw0yODExMTgyMjE5NTVaMCMxEjAQBgNVBAoTCUNvY2tyb2Fj
aDENMAsGA1UEAxMEbm9kZTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEB
AK6t6tthajEHfnmODgY4oJYLQuq8gNMW439kNTdqGn6mYNbNPtpzVqfeH8F+O24E
+bRo1GHPLZ56ujaLVX5oxOgW8d4Hi60tIXSf1ChsX+RsXAFJWHExEyIC3ljEuh/E
AqXAaD5mVjE4ba0PYvYh7QXr0CD65LHtwoZlua7eiRK8kQCIOAJDZPM8xqPq5h0c
4Te44NY8sYuknPVWOZfejMW+FPlMS4fWqe8CO1Iv9tYFg7greYKFy2wsbs7u6Hfv
+t3qA5FcmjDLyOMQZqgN46MC8uGE0IuKljAyXwgKZhuSBJxtlLSrDm03wksAilaQ
IGgV4/PyQzWisI5CMOP9MtMCAwEAAaNpMGcwDgYDVR0PAQH/BAQDAgWgMB0GA1Ud
JQQWMBQGCCsGAQUFBwMBBggrBgEFBQcDAjAfBgNVHSMEGDAWgBS9QxqanQbManN3
JRtphbZ7RfGiWTAVBgNVHREEDjAMhwR/AAABhwR/AAABMA0GCSqGSIb3DQEBCwUA
A4IBAQB/KfS2c7NrMrwJ1o09ZJe8vbODW8w2oDOiX+KcWx/Ysgublp7fK79q25Y+
jTKXwd4CnfR1XEhkALB2s/kZzUCJJQNbhpfWfsv3D7i8xFx+UVqXrX8/chcWk9pF
hYaRTz1O6SoQb31VRMIPBNp4zIgcoy7/t18YZxGQxZE2RHK/bkouHqgnImgPqr/t
BZMq4jmEySLANRUba8KRHhu2jbMfocgQoUAbob3Xu5bqurKxrSoNeqMxhNKJlZWg
yPQb3mGWDkuU3eLVUGeSmVkvKBGRek/7nCVnwuiAoap7q/FviZFOwpZj3weh7iY8
njPbh704HgY1WA3/+VI9aloX9b8h
-----END CERTIFICATE-----
```

I agree that there may be a feature request here for having the generated TLS certificate use the name specified by `--hostname` or similar.

Comment (ZhouXing19) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:116]: As @sudomateo mentioned, `--insecure` is allowed, but not the default. 

>  I figured that it was intentionally starting the cluster in secure mode even though the multi-node cluster instructions above it used --insecure. 

I wonder what extra benefit this would bring. If the motivation is to make sql console accessible locally (i.e. via `./cockroach sql` directly, rather than via `docker exec -it ./cockroach sql`), the user now is already able to do it by `docker cp` to copy the certs to local file system, and then 
```
./cockroach sql --url="postgresql://root@127.0.0.1:26257/defaultdb?sslcert=certs%2Fclient.root.crt&sslkey=certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=certs%2Fca.crt"
```

Additionally, originally `start-single-node` with docker is intended for tests, so we expect the usage of it doesn't require many customization, and default network setting should be sufficient. 

(FYI I'm currently not on the SQL Foundations team, who owns this part of the codebase, so would like to cc @dikshant and @rafiss in terms of new feature request)

Comment (sudomateo) [reply src/current/_includes/v23.1/start-in-docker/mac-linux-steps-single-node.md:116]: @mdlinville bumping this for review this week. The changes are correct in their current state but I would like to resolve this discussion before merging. Let me know what you think.

**Review by mdlinville — APPROVED** [review]
